src/providers/42.ts

```diff
@@ -0,0 +1,51 @@
+import { OAuth2Client } from "oslo/oauth2";
+
+import type { OAuth2Provider } from "../index.js";
+
+const authorizeEndpoint = "https://api.intra.42.fr/oauth/authorize";
+const tokenEndpoint = "https://api.intra.42.fr/oauth/token";
+
+export class FortyTwo implements OAuth2Provider {
+	private client: OAuth2Client;
+	private clientSecret: string;
+
+	constructor(clientId: string, clientSecret: string, redirectURI: string) {
+		this.client = new OAuth2Client(clientId, authorizeEndpoint, tokenEndpoint, {
+			redirectURI
+		});
+		this.clientSecret = clientSecret;
+	}
+
+	public async createAuthorizationURL(
+		state: string,
+		options?: {
+			scopes?: string[];
+		}
+	): Promise<URL> {
+		return await this.client.createAuthorizationURL({
+			state,
+			scopes: options?.scopes ?? []
+		});
+	}
+
+	public async validateAuthorizationCode(code: string): Promise<FortyTwoTokens> {
+		const result = await this.client.validateAuthorizationCode(code, {
+			authenticateWith: "request_body",
+			credentials: this.clientSecret
+		});
+		const tokens: FortyTwoTokens = {
+			accessToken: result.access_token,
+			refreshToken: result.refresh_token ?? null,
+			expiresIn: result.expires_in ?? null,
+			tokenType: result.token_type ?? null
+		};
+		return tokens;
+	}
+}
+
+export interface FortyTwoTokens {
+	accessToken: string;
+	refreshToken: string | null;
```

Comment:
Is `refresh_token` and `expires_in` even included in the response?

Reply:
Looks like `refresh_token` isn't included so you can just omit that, while `expires_in` is - so please replace it with `accessTokenExpiresAt` field (see other providers)